NPE when generating test report and src is null

**Citrus Version**
4.10.1

```
2026-06-23 07:52:23.209 ERROR 5183 --- [           main] usframework.report.TestFlowReporter : Failed to create test report
java.lang.NullPointerException: Cannot read the array length because "src" is null
	at java.base/java.util.Base64$Encoder.encode(Base64.java:295)
	at java.base/java.util.Base64$Encoder.encodeToString(Base64.java:349)
	at org.citrusframework.TestActionResult.toJson(TestActionResult.java:104)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.citrusframework.report.TestFlowReporter$ReportEntry.toJson(TestFlowReporter.java:328)
	at org.citrusframework.report.TestFlowReporter.lambda$generate$1(TestFlowReporter.java:79)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.citrusframework.report.TestFlowReporter.generate(TestFlowReporter.java:77)
	at org.citrusframework.report.AbstractTestReporter.generateReport(AbstractTestReporter.java:39)
	at org.citrusframework.report.TestReporters.generateReports(TestReporters.java:142)
	at org.citrusframework.report.TestReporters.onFinishSuccess(TestReporters.java:59)
	at org.citrusframework.report.TestSuiteListeners.onFinishSuccess(TestSuiteListeners.java:54)
	at org.citrusframework.Citrus.afterSuite(Citrus.java:112)
	at org.citrusframework.junit.jupiter.CitrusExtension$AfterSuiteCallback.close(CitrusExtension.java:292)
	at org.junit.jupiter.engine.descriptor.AbstractExtensionContext.lambda$createCloseAction$2(AbstractExtensionContext.java:108)
	at org.junit.platform.engine.support.store.NamespacedHierarchicalStore$EvaluatedValue.close(NamespacedHierarchicalStore.java:412)
	at org.junit.platform.engine.support.store.NamespacedHierarchicalStore$EvaluatedValue.access$800(NamespacedHierarchicalStore.java:396)
	at org.junit.platform.engine.support.store.NamespacedHierarchicalStore.lambda$close$2(NamespacedHierarchicalStore.java:139)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:261)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:261)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at org.junit.platform.engine.support.store.NamespacedHierarchicalStore.close(NamespacedHierarchicalStore.java:137)
	at org.junit.jupiter.engine.descriptor.AbstractExtensionContext.close(AbstractExtensionContext.java:126)
	at org.junit.jupiter.engine.execution.JupiterEngineExecutionContext.close(JupiterEngineExecutionContext.java:49)
	at org.junit.jupiter.engine.descriptor.JupiterEngineDescriptor.cleanUp(JupiterEngineDescriptor.java:67)
	at org.junit.jupiter.engine.descriptor.JupiterEngineDescriptor.cleanUp(JupiterEngineDescriptor.java:29)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$cleanUp$10(NodeTestTask.java:173)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.cleanUp(NodeTestTask.java:173)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:104)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.executeEngine(EngineExecutionOrchestrator.java:230)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.failOrExecuteEngine(EngineExecutionOrchestrator.java:204)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:172)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:101)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:64)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:150)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:63)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:109)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:91)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.junit.platform.launcher.core.InterceptingLauncher.lambda$execute$1(InterceptingLauncher.java:39)
	at org.junit.platform.launcher.core.ClasspathAlignmentCheckingLauncherInterceptor.intercept(ClasspathAlignmentCheckingLauncherInterceptor.java:25)
	at org.junit.platform.launcher.core.InterceptingLauncher.execute(InterceptingLauncher.java:38)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.apache.maven.surefire.junitplatform.LauncherAdapter.executeWithoutCancellationToken(LauncherAdapter.java:60)
	at org.apache.maven.surefire.junitplatform.LauncherAdapter.execute(LauncherAdapter.java:52)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:203)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:168)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:136)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter :
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : CITRUS TEST RESULTS
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter :
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 15447 ms) jira-add-issue-sink-pipe-test
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 12027 ms) jira-source-pipe-test
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 12598 ms) jira-add-comment-sink-pipe-test
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 16761 ms) aws-sqs-uri-pipe-test
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 14794 ms) aws-s3-to-http-test
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 12904 ms) aws-sqs-to-log-uri-based-test
2026-06-23 07:52:23.209  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 13843 ms) aws-s3-uri-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 15531 ms) aws-kinesis-source-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 13826 ms) aws-s3-to-log-uri-based-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 21627 ms) aws-eventbridge-sink-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 24312 ms) aws-kinesis-sink-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 13698 ms) kafka-sink-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 21081 ms) kafka-source-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : FAILURE ( 36137 ms) kafka-router-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : 	Caused by: MessageTimeoutException: Action timeout after 30000 milliseconds. Failed to receive message on endpoint: 'my-topic-3_2026-06-23'
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 20653 ms) avro-serdes-action-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 20742 ms) avro-data-type-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 12613 ms) slack-source-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 11571 ms) slack-sink-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  9269 ms) earthquake-to-http-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8433 ms) hoist-field-action-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 10771 ms) extract-field-action-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8415 ms) data-type-action-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8485 ms) insert-field-action-pipe-test
2026-06-23 07:52:23.210  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8472 ms) replace-field-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8451 ms) mask-field-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8449 ms) drop-field-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8442 ms) crypto-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8366 ms) predicate-filter-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8550 ms) has-header-filter-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8414 ms) insert-header-action-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8810 ms) log-sink-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8411 ms) timer-to-http-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 11781 ms) timer-to-http-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8499 ms) http-sink-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8960 ms) http-source-to-http-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8995 ms) rest-openapi-sink-delete-pet-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  9609 ms) rest-openapi-sink-add-pet-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS (  8832 ms) mail-sink-pipe-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 20686 ms) protobuf-data-type-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : SUCCESS ( 20442 ms) protobuf-serdes-action-test
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter :
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : TOTAL:		40
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : PASSED:		39 (97.5%)
2026-06-23 07:52:23.211  INFO 5183 --- [           main] rusframework.report.LoggingReporter : FAILED:		1 (2.5%)
2026-06-23 07:52:23.212  INFO 5183 --- [           main] rusframework.report.LoggingReporter : TIME:		529707 ms
2026-06-23 07:52:23.212  INFO 5183 --- [           main] rusframework.report.LoggingReporter :
2026-06-23 07:52:23.212  INFO 5183 --- [           main] rusframework.report.LoggingReporter : ------------------------------------------------------------------------
2026-06-23 07:52:23.216  INFO 5183 --- [           main] k.report.AbstractOutputFileReporter : Generated test report: target/citrus-reports/citrus-test-results.html
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   KafkaIT.kafka()[3] » TestCaseFailed Action timeout after 30000 milliseconds. Failed to receive message on endpoint: 'my-topic-3_2026-06-23'
[INFO]
[ERROR] Tests run: 40, Failures: 0, Errors: 1, Skipped: 0
```

--- a/core/citrus-api/src/main/java/org/citrusframework/TestActionResult.java
+++ b/core/citrus-api/src/main/java/org/citrusframework/TestActionResult.java
@@ -19,6 +19,7 @@ package org.citrusframework;
 import java.util.ArrayList;
 import java.util.Base64;
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Collectors;
 
 import org.citrusframework.json.JsonStringBuilder;
@@ -101,7 +102,9 @@ public class TestActionResult {
                                 .map(String::getBytes)
                                 .map(Base64.getEncoder()::encodeToString)
                                 .toList())
-                        .withProperty("payload", Base64.getEncoder().encodeToString(message.getPayload(byte[].class)))
+                        .withProperty("payload", Optional.ofNullable(message.getPayload(byte[].class))
+                                .map(Base64.getEncoder()::encodeToString)
+                                .orElse(""))
                     .closeObject();
         }
 

--- a/core/citrus-api/src/test/java/org/citrusframework/TestActionResultTest.java
+++ b/core/citrus-api/src/test/java/org/citrusframework/TestActionResultTest.java
@@ -1,0 +1,364 @@
+/*
+ * Copyright the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.citrusframework;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.citrusframework.message.Message;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestActionResultTest {
+
+    @Test
+    public void toJsonMinimalResult() {
+        var fixture = new TestActionResult("echo", "actions.0.echo");
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"echo\", \"path\": \"actions.0.echo\" }");
+    }
+
+    @Test
+    public void toJsonWithError() {
+        var fixture = new TestActionResult("send", "actions.1.send");
+        fixture.setError("Connection refused");
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"send\", \"path\": \"actions.1.send\", \"error\": \"Connection refused\" }");
+    }
+
+    @Test
+    public void toJsonShouldEscapeQuotesInError() {
+        var fixture = new TestActionResult("send", "actions.0.send");
+        fixture.setError("Expected \"hello\" but was \"world\"");
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"send\", \"path\": \"actions.0.send\", \"error\": \"Expected \\\"hello\\\" but was \\\"world\\\"\" }");
+    }
+
+    @Test
+    public void toJsonWithMessage() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("Content-Type", "application/json");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(List.of("<SOAP-ENV:Header/>"));
+        when(message.getPayload(byte[].class)).thenReturn("Hello".getBytes(StandardCharsets.UTF_8));
+
+        var fixture = new TestActionResult("send", "actions.1.send");
+        fixture.setMessage(message);
+
+        String headerDataBase64 = Base64.getEncoder().encodeToString("<SOAP-ENV:Header/>".getBytes());
+        String payloadBase64 = Base64.getEncoder().encodeToString("Hello".getBytes(StandardCharsets.UTF_8));
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"send\", \"path\": \"actions.1.send\", " +
+                "\"message\": { " +
+                "\"headers\": [ { \"name\": \"Content-Type\", \"value\": \"application/json\" } ], " +
+                "\"headerData\": [ \"%s\" ], ".formatted(headerDataBase64) +
+                "\"payload\": \"%s\"".formatted(payloadBase64) +
+                " } }");
+    }
+
+    @Test
+    public void toJsonWithEmptyPayload() {
+        var message = mock(Message.class);
+        when(message.getHeaders()).thenReturn(Collections.emptyMap());
+        when(message.getHeaderData()).thenReturn(Collections.emptyList());
+        when(message.getPayload(byte[].class)).thenReturn(null);
+
+        var fixture = new TestActionResult("receive", "actions.2.receive");
+        fixture.setMessage(message);
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"receive\", \"path\": \"actions.2.receive\", " +
+                "\"message\": { \"headers\": [  ], \"headerData\": [], \"payload\": \"\" } }");
+    }
+
+    @Test
+    public void toJsonWithNullPayload() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("operation", "greet");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(List.of("<Header/>"));
+        when(message.getPayload(byte[].class)).thenReturn(null);
+
+        var fixture = new TestActionResult("send", "actions.0.send");
+        fixture.setMessage(message);
+
+        String headerDataBase64 = Base64.getEncoder().encodeToString("<Header/>".getBytes());
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"send\", \"path\": \"actions.0.send\", " +
+                "\"message\": { " +
+                "\"headers\": [ { \"name\": \"operation\", \"value\": \"greet\" } ], " +
+                "\"headerData\": [ \"%s\" ], ".formatted(headerDataBase64) +
+                "\"payload\": \"\"" +
+                " } }");
+    }
+
+    @Test
+    public void toJsonWithChildActions() {
+        var fixture = new TestActionResult("sequence", "actions.0.sequence");
+        fixture.addAction(new TestActionResult("echo", "actions.0.sequence.actions.0.echo"));
+        fixture.addAction(new TestActionResult("send", "actions.0.sequence.actions.1.send"));
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"sequence\", \"path\": \"actions.0.sequence\", " +
+                "\"actions\": [ " +
+                "{ \"name\": \"echo\", \"path\": \"actions.0.sequence.actions.0.echo\" }," +
+                "{ \"name\": \"send\", \"path\": \"actions.0.sequence.actions.1.send\" }" +
+                " ] }");
+    }
+
+    @Test
+    public void toJsonWithIterations() {
+        var fixture = new TestActionResult("iterate", "actions.0.iterate");
+        fixture.addIteration(new TestActionResult("0", "actions.0.iterate"));
+        fixture.addIteration(new TestActionResult("1", "actions.0.iterate"));
+
+        Assert.assertEquals(fixture.toJson(),
+                "{ \"name\": \"iterate\", \"path\": \"actions.0.iterate\", " +
+                "\"iterations\": [ " +
+                "{ \"name\": \"0\", \"path\": \"actions.0.iterate\" }," +
+                "{ \"name\": \"1\", \"path\": \"actions.0.iterate\" }" +
+                " ] }");
+    }
+
+    @Test
+    public void toJsonWithAllFields() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("operation", "test");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(Collections.emptyList());
+        when(message.getPayload(byte[].class)).thenReturn("payload".getBytes(StandardCharsets.UTF_8));
+
+        var fixture = new TestActionResult("sequence", "actions.0.sequence");
+        fixture.setError("validation failed");
+        fixture.setMessage(message);
+        fixture.addAction(new TestActionResult("send", "actions.0.sequence.actions.0.send"));
+        fixture.addIteration(new TestActionResult("0", "actions.0.sequence"));
+
+        String json = fixture.toJson();
+
+        String payloadBase64 = Base64.getEncoder().encodeToString("payload".getBytes(StandardCharsets.UTF_8));
+
+        Assert.assertTrue(json.contains("\"name\": \"sequence\""));
+        Assert.assertTrue(json.contains("\"path\": \"actions.0.sequence\""));
+        Assert.assertTrue(json.contains("\"error\": \"validation failed\""));
+        Assert.assertTrue(json.contains("\"message\": {"));
+        Assert.assertTrue(json.contains("\"payload\": \"%s\"".formatted(payloadBase64)));
+        Assert.assertTrue(json.contains("\"actions\": ["));
+        Assert.assertTrue(json.contains("\"iterations\": ["));
+    }
+
+    @Test
+    public void toJsonWithMultipleHeaders() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("key1", "value1");
+        headers.put("key2", "value2");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(Collections.emptyList());
+        when(message.getPayload(byte[].class)).thenReturn(null);
+
+        var fixture = new TestActionResult("send", "actions.1.send");
+        fixture.setMessage(message);
+
+        String json = fixture.toJson();
+
+        Assert.assertTrue(json.contains("{ \"name\": \"key1\", \"value\": \"value1\" }"));
+        Assert.assertTrue(json.contains("{ \"name\": \"key2\", \"value\": \"value2\" }"));
+    }
+
+    @Test
+    public void toYamlMinimalResult() {
+        var fixture = new TestActionResult("echo", "actions.0.echo");
+
+        Assert.assertEquals(fixture.toYaml(),
+                "name: \"echo\"\n" +
+                "path: \"actions.0.echo\"\n");
+    }
+
+    @Test
+    public void toYamlWithError() {
+        var fixture = new TestActionResult("send", "actions.1.send");
+        fixture.setError("Connection refused");
+
+        Assert.assertEquals(fixture.toYaml(),
+                "name: \"send\"\n" +
+                "path: \"actions.1.send\"\n" +
+                "error: |\n" +
+                "  Connection refused\n");
+    }
+
+    @Test
+    public void toYamlWithMessage() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("Content-Type", "application/json");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(List.of("<SOAP-ENV:Header/>"));
+        when(message.getPayload(String.class)).thenReturn("Hello");
+
+        var fixture = new TestActionResult("send", "actions.1.send");
+        fixture.setMessage(message);
+
+        String headerDataBase64 = Base64.getEncoder().encodeToString("<SOAP-ENV:Header/>".getBytes());
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("name: \"send\""));
+        Assert.assertTrue(yaml.contains("path: \"actions.1.send\""));
+        Assert.assertTrue(yaml.contains("message:\n"));
+        Assert.assertTrue(yaml.contains("headers:\n"));
+        Assert.assertTrue(yaml.contains("- name: \"Content-Type\""));
+        Assert.assertTrue(yaml.contains("value: \"application/json\""));
+        Assert.assertTrue(yaml.contains("headerData: [ '%s' ]".formatted(headerDataBase64)));
+        Assert.assertTrue(yaml.contains("payload: |\n"));
+        Assert.assertTrue(yaml.contains("Hello"));
+    }
+
+    @Test
+    public void toYamlWithEmptyPayload() {
+        var message = mock(Message.class);
+        when(message.getHeaders()).thenReturn(Collections.emptyMap());
+        when(message.getHeaderData()).thenReturn(Collections.emptyList());
+        when(message.getPayload(String.class)).thenReturn(null);
+
+        var fixture = new TestActionResult("receive", "actions.2.receive");
+        fixture.setMessage(message);
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("name: \"receive\""));
+        Assert.assertTrue(yaml.contains("message:\n"));
+        Assert.assertTrue(yaml.contains("headerData: []"));
+        Assert.assertTrue(yaml.contains("payload: |\n"));
+    }
+
+    @Test
+    public void toYamlWithNullPayload() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("operation", "greet");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(List.of("<Header/>"));
+        when(message.getPayload(String.class)).thenReturn(null);
+
+        var fixture = new TestActionResult("send", "actions.0.send");
+        fixture.setMessage(message);
+
+        String headerDataBase64 = Base64.getEncoder().encodeToString("<Header/>".getBytes());
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("- name: \"operation\""));
+        Assert.assertTrue(yaml.contains("value: \"greet\""));
+        Assert.assertTrue(yaml.contains("headerData: [ '%s' ]".formatted(headerDataBase64)));
+        Assert.assertTrue(yaml.contains("payload: |\n"));
+    }
+
+    @Test
+    public void toYamlWithChildActions() {
+        var fixture = new TestActionResult("sequence", "actions.0.sequence");
+        fixture.addAction(new TestActionResult("echo", "actions.0.sequence.actions.0.echo"));
+        fixture.addAction(new TestActionResult("send", "actions.0.sequence.actions.1.send"));
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("name: \"sequence\""));
+        Assert.assertTrue(yaml.contains("actions:\n"));
+        Assert.assertTrue(yaml.contains("name: \"echo\""));
+        Assert.assertTrue(yaml.contains("path: \"actions.0.sequence.actions.0.echo\""));
+        Assert.assertTrue(yaml.contains("name: \"send\""));
+        Assert.assertTrue(yaml.contains("path: \"actions.0.sequence.actions.1.send\""));
+    }
+
+    @Test
+    public void toYamlWithIterations() {
+        var fixture = new TestActionResult("iterate", "actions.0.iterate");
+        fixture.addIteration(new TestActionResult("0", "actions.0.iterate"));
+        fixture.addIteration(new TestActionResult("1", "actions.0.iterate"));
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("name: \"iterate\""));
+        Assert.assertTrue(yaml.contains("iterations:\n"));
+        Assert.assertTrue(yaml.contains("name: \"0\""));
+        Assert.assertTrue(yaml.contains("name: \"1\""));
+        Assert.assertTrue(yaml.contains("path: \"actions.0.iterate\""));
+    }
+
+    @Test
+    public void toYamlWithMultipleHeaders() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("key1", "value1");
+        headers.put("key2", "value2");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(Collections.emptyList());
+        when(message.getPayload(String.class)).thenReturn(null);
+
+        var fixture = new TestActionResult("send", "actions.1.send");
+        fixture.setMessage(message);
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("- name: \"key1\""));
+        Assert.assertTrue(yaml.contains("value: \"value1\""));
+        Assert.assertTrue(yaml.contains("- name: \"key2\""));
+        Assert.assertTrue(yaml.contains("value: \"value2\""));
+    }
+
+    @Test
+    public void toYamlWithAllFields() {
+        var message = mock(Message.class);
+        Map<String, Object> headers = new LinkedHashMap<>();
+        headers.put("operation", "test");
+        when(message.getHeaders()).thenReturn(headers);
+        when(message.getHeaderData()).thenReturn(Collections.emptyList());
+        when(message.getPayload(String.class)).thenReturn("payload");
+
+        var fixture = new TestActionResult("sequence", "actions.0.sequence");
+        fixture.setError("validation failed");
+        fixture.setMessage(message);
+        fixture.addAction(new TestActionResult("send", "actions.0.sequence.actions.0.send"));
+        fixture.addIteration(new TestActionResult("0", "actions.0.sequence"));
+
+        String yaml = fixture.toYaml();
+
+        Assert.assertTrue(yaml.contains("name: \"sequence\""));
+        Assert.assertTrue(yaml.contains("path: \"actions.0.sequence\""));
+        Assert.assertTrue(yaml.contains("error: |"));
+        Assert.assertTrue(yaml.contains("validation failed"));
+        Assert.assertTrue(yaml.contains("message:\n"));
+        Assert.assertTrue(yaml.contains("payload: |"));
+        Assert.assertTrue(yaml.contains("actions:\n"));
+        Assert.assertTrue(yaml.contains("iterations:\n"));
+    }
+}

